From 33b48aaf141b8f3b33374d290e7270e9ce294ea7 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:02:02 -0400
Subject: [infiniband] mlx4: Get upper dev addresses as RoCE GIDs when port comes up

Message-id: <161ed243b07cc629278b36c38d2e2faec31c83a5.1410425016.git.avadai@redhat.com>
Patchwork-id: 93722
O-Subject: [Patch RHEL7.1 V1 186/197] IB/mlx4: Get upper dev addresses as RoCE GIDs when port comes up
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Moni Shoua <monis@mellanox.com>

When RoCE port becomes active and netdev of the port has upper
device (e.g bond/team), gids derived from the upper dev should
appear in the port's RoCE gid table.

Signed-off-by: Moni Shoua <monis@mellanox.com>
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>

Upstream: bccb84f1dfab92ed180adf09c76cfa9ddc90edb9
Signed-off-by: Amir Vadai <avadai@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/mlx4/main.c b/drivers/infiniband/hw/mlx4/main.c
index 54c2b3a..0879ee8 100644
--- a/drivers/infiniband/hw/mlx4/main.c
+++ b/drivers/infiniband/hw/mlx4/main.c
@@ -1799,14 +1799,23 @@ static void mlx4_ib_scan_netdevs(struct mlx4_ib_dev *ibdev,
    port_state = (netif_running(curr_netdev) && netif_carrier_ok(curr_netdev)) ?
       IB_PORT_ACTIVE : IB_PORT_DOWN;
    mlx4_ib_set_default_gid(ibdev, curr_netdev, port);
-   /* if using bonding/team and a slave port is down, we
-    * don't the bond IP based gids in the table since
-    * flows that select port by gid may get the down port.
-    */
-   if (curr_master && (port_state == IB_PORT_DOWN)) {
-    reset_gid_table(ibdev, port);
-    mlx4_ib_set_default_gid(ibdev,
-       curr_netdev, port);
+   if (curr_master) {
+    /* if using bonding/team and a slave port is down, we
+     * don't want the bond IP based gids in the table since
+     * flows that select port by gid may get the down port.
+    */
+    if (port_state == IB_PORT_DOWN) {
+     reset_gid_table(ibdev, port);
+     mlx4_ib_set_default_gid(ibdev,
+        curr_netdev,
+        port);
+    } else {
+     /* gids from the upper dev (bond/team)
+      * should appear in port's gid table
+     */
+     mlx4_ib_get_dev_addr(curr_master,
+            ibdev, port);
+    }
    }
    /* if bonding is used it is possible that we add it to
     * masters only after IP address is assigned to the
-- 
1.7.1